tests/test_client.py

```diff
@@ -422,6 +422,100 @@ async def iterator_func(sub):
         self.assertEqual("tests.3", msgs[3].subject)
         await nc.close()
 
+    @async_test
+    async def test_subscribe_iterate_unsub_comprehension(self):
+        nc = NATS()
+        msgs = []
+
+        await nc.connect()
+
+        # Make subscription that only expects a couple of messages.
+        sub = await nc.subscribe('tests.>')
+        await sub.unsubscribe(limit=2)
+        await nc.flush()
+
+        for i in range(0, 5):
+            await nc.publish(f"tests.{i}", b'bar')
+
+        # A couple of messages would be received then this will unblock.
+        msgs = [msg async for msg in sub.messages]
+
+        self.assertEqual(2, len(msgs))
+        self.assertEqual("tests.0", msgs[0].subject)
+        self.assertEqual("tests.1", msgs[1].subject)
+        await nc.drain()
+
+    @async_test
+    async def test_subscribe_iterate_unsub(self):
+        nc = NATS()
+        msgs = []
+
+        await nc.connect()
+
+        # Make subscription that only expects a couple of messages.
+        sub = await nc.subscribe('tests.>')
+        await sub.unsubscribe(limit=2)
+        await nc.flush()
+
+        for i in range(0, 5):
+            await nc.publish(f"tests.{i}", b'bar')
+
+        # A couple of messages would be received then this will unblock.
+        msgs = []
+        async for msg in sub.messages:
+            msgs.append(msg)
+
+            if len(msgs) == 2:
+                await sub.unsubscribe()
+
+        self.assertEqual(2, len(msgs))
+        self.assertEqual("tests.0", msgs[0].subject)
+        self.assertEqual("tests.1", msgs[1].subject)
+        await nc.drain()
+
+    @async_test
+    async def test_subscribe_iterate_next_msg(self):
+        nc = NATS()
+        msgs = []
+
+        await nc.connect()
+
+        # Make subscription that only expects a couple of messages.
+        sub = await nc.subscribe('tests.>')
+        await nc.flush()
+
+        # Simple iterator to get next msg.
+        async def next_msg():
+            async for msg in sub.messages:
+                return msg
```

Comment:
Think it would be cool to have a `sub.next_msg` like in the Go client as well, but if the future here times out then the message is lost, not sure right now but maybe there is a better way...

Reply:
The future could potentially be stored on the subscription along with a flag indicating whether the result has been retrieved or not (if not, return/await on the future result, if not, create a new future and return/await on that)